### PR TITLE
Calculate push actions on worker

### PR DESCRIPTION
**Pull request by erikjohnston**
Includes #2874

### CODE REVIEW

**Comment by erikjohnston** [comment]
This includes #2913

File: synapse/handlers/message.py [hunk 1]
```diff
@@ -553,21 +553,67 @@ def handle_new_client_event(
...
+    @defer.inlineCallbacks
+    def persist_and_notify_client_event(
+        self,
+        requester,
+        event,
+        context,
+        ratelimit=True,
+        extra_users=[],
+    ):
+        """Called when we have fully built and authed the event. This should
```
Comment (richvdh) [inline synapse/handlers/message.py]: handle_new_client_event claims "This includes checking auth", but then doesn't seem to do any auth checks before calling this, which says "Called when we have ... authed the event". Something seems to be lying.

Perhaps you could also mention push actions here and how they are handled.

File: synapse/handlers/message.py [hunk 2]
```diff
@@ -553,21 +553,67 @@ def handle_new_client_event(
...
+        notifying users, sending to remote servers, etc.
+
+        If called from a worker will hit out to the master process for final
+        processing.
+
+        Args:
+            requester (Requester)
+            event (FrozenEvent)
+            context (EventContext)
+            ratelimit (bool)
```
Comment (richvdh) [inline synapse/handlers/message.py]: we don't seem to do  much with this when we have a worker.

File: synapse/handlers/message.py [hunk 3]
```diff
@@ -553,21 +553,67 @@ def handle_new_client_event(
...
+            extra_users (list(str)): Any extra users to notify about event
+        """
+
+        yield self.action_generator.handle_push_actions_for_event(
+            event, context
+        )
+
+        try:
+            # We now need to go and hit out to wherever we need to hit out to.
```
Comment (richvdh) [inline synapse/handlers/message.py]: this is a remarkably uninformative comment. Suggest ditching it.

File: synapse/replication/http/send_event.py [hunk 4]
```diff
@@ -106,7 +106,7 @@ def on_POST(self, request):
             event.event_id, event.room_id,
         )
 
-        yield self.event_creation_handler.handle_new_client_event(
+        yield self.event_creation_handler.persist_and_notify_client_event(
```
Comment (richvdh) [inline synapse/replication/http/send_event.py]: do we not need to remove the push actions from the staging area if this fails?

Comment (erikjohnston) [reply synapse/replication/http/send_event.py]: We remove push actions in a try finally block after we call `handle_push_actions_for_event` in `handle_new_client_event`

File: synapse/handlers/message.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/message.py]: Woops, must haven't confused with another PR where we hoist all the auth stuff to the worker too.

File: synapse/handlers/message.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/message.py]: Ah, and ditto for the `extra_users` param. I've added a commit to send the two params over replication.

File: synapse/replication/http/send_event.py (hunk 4 above)
Comment (richvdh) [reply synapse/replication/http/send_event.py]: Ah yes. I was forgetting that the call to the master was synchronous and that the error would propagate back to the worker, which will do the cleanup.

**Comment by richvdh** [comment]
As discussed in synapse-dev, we probably need to worry about what happens if the http hit to the master times out, but this looks ok so far.

**Pull request merged**